### PR TITLE
Exposing a change_no_check tactic

### PR DESCRIPTION
**Pull request by herbelin**
**Kind:** feature

This was suggested by #10012, as a possible "natural" user-level entry point.

Would need to be synchronized with #9983 [Done]

- [X] Added / updated test-suite

- [ ] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [X] Entry added in CHANGES.md.

### CODE REVIEW

File: plugins/ltac/g_tactic.mlg [hunk 1]
```diff
@@ -703,7 +703,11 @@ GRAMMAR EXTEND Gram
       | IDENT "change"; c = conversion; cl = clause_dft_concl ->
 	  { let (oc, c) = c in
 	  let p,cl = merge_occurrences loc cl oc in
-	  TacAtom (CAst.make ~loc @@ TacChange (p,c,cl)) }
+          TacAtom (CAst.make ~loc @@ TacChange (true,p,c,cl)) }
```
Comment (Zimmi48) [inline plugins/ltac/g_tactic.mlg]: Could you take advantage of this change to translate the French comment just above?

File: tactics/tactics.ml [hunk 2]
```diff
@@ -819,15 +819,21 @@ let change_and_check cv_pb mayneedglobalcheck deep t env sigma c =
...
+        t Id.Map.empty env sigma
   | Some occl ->
       e_contextually false occl
         (fun subst ->
-          change_and_check Reduction.CONV mayneedglobalcheck true (t subst))
-        env sigma c in
+          if check then
+            change_and_check Reduction.CONV mayneedglobalcheck true (t subst)
+          else
+            fun env sigma c -> t subst env sigma) env sigma c in
```
Comment (Zimmi48) [inline tactics/tactics.ml]: This location for the arguments `env sigma c` looks quite strange to me.

**Review by Zimmi48 — COMMENTED** [review]
I think this new tactic would deserve a test file in the test-suite.

Could you also clarify the distinction with `convert_concl_no_check`? I thought the point was just to rename it but it seems that the implementation is different, *and* you don't deprecate `convert_concl_no_check` as part of this PR.

**Comment by herbelin** [comment]
> I think this new tactic would deserve a test file in the test-suite.

Yes. I'll do it eventually when it is clear that the direction taken by this PS is validated.

> Could you also clarify the distinction with `convert_concl_no_check`? I thought the point was just to rename it but it seems that the implementation is different, and you don't deprecate `convert_concl_no_check` as part of this PR.

`change_no_check` is more general than `convert_concl_no_check`. In addition to allowing changing the conclusion, it has all the other features of `change`, such as `change foo in hyp`, or `change bla with bli`. 

If giving a focus on `change_no_check` is validated, `convert_concl_no_check` can then indeed be deprecated.

**Comment by Zimmi48** [comment]
Thanks, then let's see what @ppedrot says.

**Comment by ppedrot** [comment]
I'm in favour of the renaming + deprecation.

File: tactics/tactics.ml (hunk 2 above)
Comment (herbelin) [reply tactics/tactics.ml]: I don't understand what you mean. Which location are you talking about?

**Comment by herbelin** [comment]
I added a mini-test and a deprecation (I followed the model of deprecation of `autoapply_using`). [Next step could be to have a generic `no_check` modifier??]

**Comment by ppedrot** [comment]
I don't like too much the idea of a no_check modifier. These tactics are somewhat unsafe, they should rather be in their own namespace if anything.

Also, the CI looks broken.

**Comment by herbelin** [comment]
CI: was a non aborted pending proof. Fixed.

A no_check modifier: that was of course to avoid proliferation of syntax. Having a namespace looks good (if ever we need no_check tactics for more than conversion).

What to do with the documentation? #10012 was going to document `convert_concl_no_concl`. Could it be just adapted to document `change_no_concl` instead?

**Comment by Zimmi48** [comment]
> What to do with the documentation? #10012 was going to document `convert_concl_no_concl`. Could it be just adapted to document `change_no_concl` instead?

I think so, yes. I would go ahead and merge this PR without documentation then.

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
Do I understand correctly that this is ready to merge?

**Comment by Zimmi48** [comment]
I think so. Maybe add a CHANGES entry though.

**Comment by ppedrot** [comment]
Isn't that the point of #10012?

**Comment by ppedrot** [comment]
All right, I think I'll merge now and postpone the CHANGE diff to #10012, otherwise I don't understand why we don't merge both PRs into a single one.

**Comment by Zimmi48** [comment]
OK fine.

**Pull request merged**